### PR TITLE
buildRustCrate: Allow version specific dependency renaming + parameter doc

### PR DESCRIPTION
**Pull request by kolloch**
This PR allows version specific crate renames:
   
```nix
pkgs.buildRustCrate {
  # ...
  crateRenames = {
     my_crate_name = [
         { version = "1.2.3"; rename = "my_alternative_name01" ; }
         { version = "3.2.3"; rename = "my_alternative_name03" ; }
     ];
     # ...
   };
```

I also documented all buildRustCrate parameters.

###### Motivation for this change
   
This roughly corresponds to the following snippet in Cargo.toml:
   
   ```toml
   [dependencies]
   my_alternative_name01 = { package = "my_crate_name", version = "0.1" }
   my_alternative_name03 = { package = "my_crate_name", version = "0.3" }
   ```
   
 Including multiple versions of a crate is very popular during   ecosystem transitions, e.g. from futures 0.1 to futures 0.3.

Corresponding issue in crate2nix: https://github.com/kolloch/crate2nix/issues/83

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s) `nix build -f . buildRustCrateTests`
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions 
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests)) (See above)
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/rust/build-rust-crate/default.nix [hunk 1]
```diff
@@ -42,11 +42,95 @@ let
...
+   # A list of extra options to pass to rustc.
+   #
+   # Example: [ "-Z debuginfo=2" ]
+   # Default: []
+   , extraRustcOpts
+   # Whether to enable building tests.
+   # Use true to enable.
+   # Default: false
+   , buildTests
+   # Passed to stenv.mkDerivation.
```
Comment (andir) [inline pkgs/build-support/rust/build-rust-crate/default.nix:160]: Missing a bunch of `d`'s in the `stdenv`.

File: pkgs/build-support/rust/build-rust-crate/default.nix [hunk 2]
```diff
@@ -172,6 +221,7 @@ stdenv.mkDerivation (rec {
 
     src = crate.src or (fetchCrate { inherit (crate) crateName version sha256; });
     name = "rust_${crate.crateName}-${crate.version}${lib.optionalString buildTests_ "-test"}";
+    version = crate.version or null;
```
Comment (andir) [inline pkgs/build-support/rust/build-rust-crate/default.nix:224]: Is the `or null` case actually required? Do we have crates without a version? Looking one line above we are already using the version as part of the `name` attribute. I would assume it is not optional.

**Review by andir — CHANGES_REQUESTED** [review]
Two minor remarks. Overall this looks good.

I also see the need for version specific overrides for things like (native)buildInputs.

Currently you have to "switch" on the version or name attribute in the overrides if you need different native inputs for the same crate name.

We should probably consider what larger refactoring we can do additionally to ease such pains.

File: pkgs/build-support/rust/build-rust-crate/default.nix (hunk 1 above)
Comment (kolloch) [reply pkgs/build-support/rust/build-rust-crate/default.nix]: Thx for catching this.

File: pkgs/build-support/rust/build-rust-crate/default.nix (hunk 2 above)
Comment (kolloch) [reply pkgs/build-support/rust/build-rust-crate/default.nix]: I guess not. I wanted to be defensive and did not see that I could already reasonably expect the version to be required.

**Comment by kolloch** [comment]
Thank you for the review, @andir !

For buildInputs, the crateOverrides mechanism can be used. This is a bit special because it concerns the extern name with which a certain crate is used within a dependent crate. So maybe those separate concerns can be handled better but will still need to be handled differently.

**Pull request merged**